text/0003-placeholders-in-conditions.md

```diff
@@ -0,0 +1,141 @@
+
+2. Template-based policy construction prevents code injection attacks (i.e., in the style of SQL injection). Templates act like prepared statements, where placeholders can only be instantiated with data values; they cannot be instantiated with strings that could be reinterpreted as arbitrary policy code. Using templates whenever possible is an important protection for applications that add policies as a result of user actions.
+
+This RFC was motivated by a particular customer use-case, i.e., [#81](https://github.com/cedar-policy/cedar/issues/81).
+
+# Detailed design
+
+This RFC does not propose any change to what syntax is valid in the policy scope. For instance, the following remains _invalid_:
+```
+permit(
+  principal,
+  action,
+  ?resource in resource
+);
+```
+The only valid uses of `?resource` in the policy scope are `resource == ?resource` and `resource in ?resource`, and this RFC is not proposing to change that. This RFC also disallows the use of `?principal` and `?resource` in the condition if they don't _also_ appear in the scope. 
+
+Both restrictions are in place because these two placeholder variables have special meaning for indexing policies in a policy store. In particular, looking _only_ at the linking information (i.e., to what `?principal` and/or `?resource` are bound) is sufficient to precisely select linked policies for particular requests. If the variables could only be bound in the condition or in other positions in the scope, and used in arbitrary contexts, it would lead to selecting superfluous policies. 
+
+Given these restrictions, the above policy would be impossible to write without support for placeholder variables beyond `?resource` and `?principal`. With their support, we can write: 
+```
+template (?lowerbound)
+permit(principal, action, resource)
+when { ?lowerbound in resource };
+```
+Here, the `template (?lowerbound)` part _introduces_ a fresh placeholder variable which can be used legally within the policy template that follows. Eliding the `template (?lowerbound)` prefix would be illegal, signaling an unbound variable. 
+
+Multiple variables can be introduced together, separated by commas, e.g.,
+```
+template (?user, ?resourcebound)
+permit(principal, action, resource)
+when { 
+     ?resourcebound in resource 
+  && principal in ?user.delegates
+};
+```
+
+All introduced placeholder variables presented so far are assumed to be entities (just like `?principal` and `?resource`). The particular _type_ of the entity is not given, for simplicity. Eliding the type poses a problem for validating templates on their own; we would have to wait to validate a template each time it is linked. However, link-time validation makes it straightforward to allow placeholder variables to have any type (`string`, `bool`, etc.), not just entity types. 
```

Comment:
This is not quite clear: is this RFC proposing that the new variables must be entities, or that they can be any type (`string`, `bool`, etc)?  Where does this RFC stop and future speculation begin?  If the part about any type is future speculation, it might be clearer to move that part out of the "Detailed design" section and into another section ("Alternatives" or a dedicated "Possible future extensions" section).

Reply:
Not future speculation -- my intention is to support any type, now. Updated line 21 for further clarification.